master/internal/kubernetes/tokens.go

```diff
@@ -0,0 +1,125 @@
+//
+// There are two reasons the token system is required as opposed to allowing the pod actors
+// to create Kubernetes resources asynchronously:
+//
+//    1) Each pod creation first requires the creation of a configMap, however creating the two
+//       is not an atomic operation. If there is a large number of concurrent creation requests
+//       (e.g., a large HP search experiment) the kubernetes API server ends up processing the
+//       creation of all the configMaps before starting to create pods, which adds significant
+//       latency to the creation of pods.
+//
+//    2) If all creation requests are submitted asynchronously, it is possible the Kubernetes API
+//       server will temporarily become saturated, and be slower to respond to other requests.
+//
+//  The reason that we implemented a token system rather than creating a worker pool which
+//  processes the creation requests themselves (the way deletion is done), is to make it possible
+//  to cancel creation requests after they are created, but before they are executed. Since the
+//  actor system processes messages in a FIFO order, any cancellation request would only be
+//  processed after the creation request case already been processed, requiring an unnecessary
+//  resource creation and deletion. An example of this is when a large HP search is created and
+//  then killed moments later. By having the pod actors request tokens, if the termination request
+//  arrives while the pod actor is awaiting to receive a token, the pod actor can just cancel its
+//  request for a token and avoid an unnecessary pod creation and deletion.
+
+const (
+	numTokens     = 5
+	tokenCoolDown = time.Millisecond * 250
+)
+
+// message types received by the tokens actor.
+type (
+	requestToken struct {
+		handler *actor.Ref
+	}
+
+	releaseToken struct {
+		handler *actor.Ref
+	}
+)
+
+// message types sent by the tokens actor.
+type (
+	grantToken struct{}
+)
+
+type tokens struct {
+	provisionedTokens map[*actor.Ref]bool
+	cancelledTokens   map[*actor.Ref]bool
+}
+
+func newTokens() *tokens {
+	return &tokens{
+		provisionedTokens: make(map[*actor.Ref]bool),
+		cancelledTokens:   make(map[*actor.Ref]bool),
+	}
+}
+
+// Receive implements the actor interface.
+func (t *tokens) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+
+	case requestToken:
+		if err := t.receiveAcquireToken(ctx, msg); err != nil {
+			return err
+		}
+
+	case releaseToken:
+		t.receiveReleaseToken(ctx, msg)
+
+	default:
+		ctx.Log().Errorf("unexpected message %T", msg)
+		return actor.ErrUnexpectedMessage(ctx)
+	}
+
+	return nil
+}
+
+func (t *tokens) receiveAcquireToken(ctx *actor.Context, msg requestToken) error {
+	if _, handlerPresent := t.cancelledTokens[msg.handler]; handlerPresent {
+		delete(t.cancelledTokens, msg.handler)
+		return nil
+	}
+
+	if len(t.provisionedTokens) == numTokens {
+		actors.NotifyAfter(ctx, tokenCoolDown, msg)
```

Comment:
question: Why not just respond with a token or a failure immediately to the pod actor?

Reply:
blocking: I understand we have limited the concurrent requests so that the k8s server won't be overwhelmed. I think using `notifyAfter` doesn't really solve all the issues. It's possible that some pod actors just can't get the tokens and keep retrying until the tokens actor finishes processing all other messages. If there is a limit on the k8s cluster, users would just see some experiments never get scheduled unless they stop all other experiments. This is caused by us using the actor message inbox as the queue to handle token requests while you can't delete messages from that inbox. A solution might be using a separate queue to handle all the requests, which shouldn't too complicated.

Reply:
Yeah we do lose FIFO here, otherwise I don't think its a huge issue as pod creation time is <<< trial run time. I realized when talking to @eecsliu yesterday that it is possible to use a worker queue instead of using tokens (requires refactoring how the worker queue for deletion is currently implemented). Going to give it a shot today and see how it turns out.